### PR TITLE
Mck/vld

### PR DESCRIPTION
**Pull request by mikolaichuk**
## Issue
https://app.clickup.com/t/2yk436u

## Description
Added new impl Validate structures. Validation added in suitable places of
lending pallet. BoundedVec is used in liquidate().

### CODE REVIEW

**Review by kevin-composable — DISMISSED** [review]
Everything looks great Mikhail

**Review by benluelo — COMMENTED** [review]
Looks good to me, just a few small things

File: frame/lending/src/benchmarking.rs [hunk 1]
```diff
@@ -104,7 +107,7 @@ benchmarks! {
 		let amount: BalanceOf<T> = 1_000_000_u64.into();
 
 		let market_id = create_market_from_raw_origin::<T>(origin.clone(), input);
-	}: _(origin, market_id, amount, false)
+	}: _(origin, market_id, Validated::new(amount).unwrap(), false)
```
[the remark this answers is not in the document]
Comment (mikolaichuk) [reply frame/lending/src/benchmarking.rs]: Fixed

File: frame/lending/src/tests.rs [hunk 2]
```diff
@@ -1523,23 +1560,18 @@ fn current_interest_rate_test() {
 }
 
 #[test]
+#[should_panic]
```
[the remark this answers is not in the document]
Comment (mikolaichuk) [reply frame/lending/src/tests.rs]: Fixed

File: frame/composable-traits/src/lending/mod.rs [hunk 3]
```diff
@@ -130,6 +159,18 @@ impl<LiquidationStrategyId, AssetId: Copy, BlockNumber>
 	}
 }
 
+#[derive(RuntimeDebug, PartialEq, TypeInfo, Default, Copy, Clone)]
+pub struct BalanceGreaterThenZero;
+impl<B> Validate<B, BalanceGreaterThenZero> for BalanceGreaterThenZero
+where
+	B: Zero + PartialOrd,
+{
+	fn validate(ballance: B) -> Result<B, &'static str> {
```
[the remark this answers is not in the document]
Comment (mikolaichuk) [reply frame/composable-traits/src/lending/mod.rs]: Fixed

File: frame/lending/src/lib.rs [hunk 4]
```diff
@@ -627,12 +631,14 @@ pub mod pallet {
...
+			input: Validated<
+				CreateInputOf<T>,
+				(MarketModelValid, CurrencyPairIsNotSame, AssetIsSupportedByOracle<T::Oracle>),
+			>,
 			keep_alive: bool,
 		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			let input = input.value();
-			let pair = input.currency_pair;
+			let pair = input.clone().value().currency_pair;
```
Comment (dzmitry-lahoda) [inline frame/lending/src/lib.rs:641]: not sure, it was working without clone. not big deal for create, but still can avoid clone?

**Review by dzmitry-lahoda — APPROVED** [review]

**Comment by dzmitry-lahoda** [comment]
also I (and some others) prefer `Oracle::`, not `O::` ingenerics.

**Review by benluelo — APPROVED** [review]

**Review by dzmitry-lahoda — APPROVED** [review]

**Comment by mikolaichuk** [comment]
bors merge

**Pull request merged**